soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -207,12 +219,123 @@ static int dt_init(const struct device *dev)
 	return intel_adsp_ipc_init(dev);
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->resume_fn = fn;
+	api->resume_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->suspend_fn = fn;
+	api->suspend_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks.
+ *
+ * @note Power state of the IPC Device is changed atomically.
+ * PM lock is taken at the start of each power transition to prevent power state change
+ * if the process fails.
+ * If IPC Device performs hardware operation and device is busy (what should not happen)
+ * function returns failure. It is API user responsibility to make sure we are not entering
+ * device power transition while device is busy.
+ */
+static int ipc_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	if (pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV))
+		return -EBUSY;
+
+	const struct ipc_control_driver_api *api =
+		(const struct ipc_control_driver_api *)dev->api;
+
+	int ret = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_state_lock(dev);
+		if (api->suspend_fn) {
+			ret = api->suspend_fn(dev, api->suspend_fn_args);
+			if (ret)
+				break;
```

Comment:
What happens here ? state will continue locked ? The system won't be able to recover from it since pm_device_action_run checks if the device's state is locked before trigger the action callback.